examples/loki/nurf_data.py

```diff
@@ -102,120 +102,173 @@ def nurf_file_creator(loki_file, path_to_loki_file, data):
 
     # open the file and append
     with h5py.File(loki_file, 'a') as hf:
+                
+        #comment on names
+        #UV/FL_Background is the dark
+        #UV/FL_Intensity0 is the reference
+        #UV/FL_Spectra is the sample
         
+        # image_key: number of frames (nFrames) given indirectly as part of the shape of the arrays 
+        # TODO: keep in mind what happens if multiple dark or reference frames are taken
+        
+        #I need to reshape data['UV_spectra']
+        data['UV_spectra']=np.reshape(data['UV_spectra'],(np.shape(data['UV_spectra'])[0],np.shape(data['UV_spectra'])[1]))
+    
+        # assemble all spectra in one variable
+        uv_all_data=np.column_stack((data['UV_spectra'].T,data['UV_background'], data['UV_intensity0']))
```

Comment:
Is this the desired data layout? If I understand correctly then `wavelength` is now the outer dimension (axis 0) of `uv_all_data`. That is, loading a single spectrum from the file will be very inefficient, but loading a specific wavelength for all spectra is efficient. Is this intentional?

Reply:
Note that this will impact in particular *appending* new spectra. With this layout you have to rewrite the entire dataset, I think, so this becomes very problematic.

Reply:
> Is this the desired data layout? If I understand correctly then `wavelength` is now the outer dimension (axis 0) of `uv_all_data`. 
Yes, the `wavelength` is axis 0. I would think once we have the setup in the lab ready and we have decided how to interact with it, we will know.  
> That is, loading a single spectrum from the file will be very inefficient, but loading a specific wavelength for all spectra is efficient. Is this intentional?
I am sorry, but I don't understand this. If there is incoming data at different points in time, we have to load the new spectrum and with it the wavelength. Probably, the wavelength range won't change in such a case.

Reply:
> Note that this will impact in particular _appending_ new spectra. With this layout you have to rewrite the entire dataset, I think, so this becomes very problematic.

What would be your suggestion please? I understood all data has to be stored as an array with the `image_key` telling us about the type of spectrum.

Reply:
> Yes, the wavelength is axis 0. I would think once we have the setup in the lab ready and we have decided how to interact with it, we will know.

I feel that this is wrong. Axis 0 means that it will be the *slowest* index (in memory/on disk), i.e., to plot say a single spectrum out of a stack of 50 spectra, you need to load values from index 0, 50, 100, 150, ...

> I am sorry, but I don't understand this. If there is incoming data at different points in time, we have to load the new spectrum and with it the wavelength. Probably, the wavelength range won't change in such a case.

This is not about changing the wavelength range. The points is that it will not be possible to append to the "stack" of spectra if wavelength is the outer dimension (axis 0).

> What would be your suggestion please? I understood all data has to be stored as an array with the image_key telling us about the type of spectrum.

The spectrum should be axis 0, wavelength should be axis 1.

Reply:
Could we discuss this in a meeting, please? I thought, all spectra relevant to one sample (sample, dark, reference) should be in one dataset as one data array. That is what I have done here. 
In a static measurement, I generate first dark, reference, then sample. If it is a time series, I would insert the new spectra at the correct position. 

I have never heard about axis 0 or axis 1. Sorry.
When I look at the example in https://manual.nexusformat.org/classes/base_classes/NXdata.html it says to plot something against time, also pressure as function of time. Then it is written `@axes=["time", "pressure"]`. For me time  is in 0 position.

Reply:
> I thought, all spectra relevant to one sample (sample, dark, reference) should be in one dataset as one data array. That is what I have done here.

Correct.

> If it is a time series, I would insert the new spectra at the correct position.

Correct. But in the current layout adding new spectra to the same file later is difficult.

> I have never heard about axis 0 or axis 1

This is just the order of indices. All I am saying is that you have transposed your data but probably should not. See for example https://numpy.org/doc/stable/glossary.html#term-axis.

> Then it is written @axes=["time", "pressure"]. For me time is in 0 position.

Correct. Time in this example is "axis 0". I think "time" for you are the different spectra you record (but there are no values for it, so you use `"."` in the Nexus attribute) and instead of a "pressure" axis you have "wavelength".

Reply:
I found `[](https://manual.nexusformat.org/datarules.html?highlight=fast%20axis#nexus-array-storage-order)NeXus Array Storage Order
NeXus stores multi-dimensional arrays of physical values in C language storage order, where the first dimension has the slowest varying index when iterating through the array in storage order, and the last dimension is the fastest varying. This is the rule. Good reasons are required to deviate from this rule.`
But I have to learn what it means.

Reply:
This is the same as the numpy default. It means that you most likely want to swap the order. This will possibly also simplify your code for combining dark, reference, and sample.

Reply:
>Correct. Time in this example is "axis 0". I think "time" for you are the different spectra you record (but there are no values for it, so you use "." in the Nexus attribute) and instead of a "pressure" axis you have "wavelength".
Maybe here is our misunderstanding ? A spectrum is `f(lambda)`. f is the recorded spectrum, lambda  as wavelength on the x-axis. In the given example, I would plot pressure as function of time, i.e. p(t). Time is on the x-axis, then it is axis 0, first axis. Pressure varies with time, and it is the second axis, axis 1.
If I do a static experiment it would be f(lambda). If there is a time-dependence it would be f(lambda, t).

Reply:
>Correct. But in the current layout adding new spectra to the same file later is difficult.

But isn't it due to the fact that we have now an `image_key` with example `[0,0,0,0,2,4]`. Are we allowed to reverse the order to make it easier? `[2,4,0,0,0,0]`?
I thought the `image_key` would make insertions easier, because it is like an index.

Reply:
> If there is a time-dependence it would be f(lambda, t).

Sure, but you are mostly interested in f(lambda) at a specific t, i.e., you rarely plot f(t) at a specific lambda? Therefore, you should not transpose the array of values, and make lambda the "fast index", "inner dimension", or axis 1 (in the 2d case).

> Are we allowed to reverse the order to make it easier? [2,4,0,0,0,0]?

Why would you reverse it? The discussion is about swapping the axes of the values array, not about reversing one of the axes.

> I thought the image_key would make insertions easier, because it is like an index.

Yes, but not if you use the wrong order of axes in your values array. If you make "wavelength" the inner dimension (axis 1) then inserting a new spectrum means appending a new row (HDF5 can do this, without rewriting the whole dataset). But as it is stored now, to insert a new spectrum you need to (1) extract all the columns (2) append a new value to each column (3) put all columns back together (4) re-write the whole dataset. This is really inefficient an messy.